### PR TITLE
feat(training): domain-adapted embeddings via iterative hard-negative mining

### LINKED ISSUE #1131: Implement Domain-Aware Embedding Service via LLM Distillation
## Overview

Implement domain-aware embedding service for semantic-router based on the paper [Adaptation of Embedding Models to Financial Filings via LLM Distillation](https://arxiv.org/pdf/2512.08088).

**Goal**: Provide an embedding API (like OpenAI's) with competitive advantages:
1. **Domain-aware embeddings** - Specialized for specific domains via iterative hard-negative mining
2. **Long context support** - Leverage existing models with 32K context

## Business Value

Enable semantic-router to serve as an embedding provider for RAG systems and agents, competing with OpenAI's embedding API through domain specialization.

---

## ✅ POC Complete: Medical Domain

**Result: +71.18% MRR@5 improvement** (exceeded +15-25% target)

| Metric | Baseline | After Training | Improvement |
|--------|----------|----------------|-------------|
| **MRR@5** | 0.4354 | 0.7453 | **+71.18%** |
| **Recall@5** | 0.4896 | 0.6900 | **+40.92%** |

### Pre-trained Model

The trained medical embedding model is available on HuggingFace:
**[llm-semantic-router/mmbert-embed-medical](https://huggingface.co/llm-semantic-router/mmbert-embed-medical)**

```python
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("llm-semantic-router/mmbert-embed-medical", trust_remote_code=True)
embeddings = model.encode(["What are the symptoms of diabetes?"])
```

### Implementation: `src/training/domain_adapted_embeddings/`

| File | Purpose |
|------|---------|
| `prepare_data.py` | Converts Q&A datasets to training format |
| `train.py` | Iterative hard-negative mining + fine-tuning |
| `README.md` | Overview and results |
| `USAGE.md` | Training instructions and code examples |

### How It Works

**The Core Idea**: Fine-tune an embedding model so questions and their correct answers have closer vector representations.

**Process**:
1. **Encode** all questions and answers into vectors using current model
2. **Rank** answers for each question by similarity
3. **Find hard cases**:
   - Hard positives: correct answers ranked too low
   - Hard negatives: wrong answers ranked too high
4. **Create triplets**: (question, correct_answer, wrong_answer)
5. **Train**: Adjust model to pull correct answers closer, push wrong answers away
6. **Repeat**: Mine new hard cases with improved model, accumulate triplets

```
Iteration 1: Mine → Train → Evaluate (MRR: 0.4354 → 0.6528, +49.93%)
Iteration 2: Mine → Accumulate → Train → Evaluate (MRR: 0.6528 → 0.7453, +71.18%)
```

### Adaptation from Paper

| Paper Approach | Our Implementation | Why |
|----------------|-------------------|-----|
| LLM-as-judge scores relevance (1-4) | Ground-truth labels | We had labeled Q&A data, no LLM needed |
| Generate queries from corpus | Use existing questions | MedQuAD provides questions |
| Chunk large PDFs | Use answers directly | Q&A answers are already reasonable size |

The paper's core contribution (iterative hard-negative mining) is preserved. We simplified by using ground-truth labels instead of LLM scoring — this works when you have labeled Q&A data.

### Critical Hyperparameters

| Parameter | Value | Why It Matters |
|-----------|-------|----------------|
| Learning rate | 5e-5 | Lower values (5e-7) result in minimal improvement |
| Margin | 0.1 | Default (~5.0) causes severe forgetting |
| Easy:Hard ratio | 2:1 | Easy triplets reinforce existing knowledge |
| Accumulate triplets | Yes | Don't replace, add each iteration |

### Training Configuration

| Parameter | Value |
|-----------|-------|
| Base model | llm-semantic-router/mmbert-embed-32k-2d-matryoshka |
| Dataset | MedQuAD (13,125 training queries) |
| Iterations | 2 |
| Epochs per iteration | 2 |
| Batch size | 8 |
| GPU | NVIDIA L4 (24GB) |

---

## Dataset Format

Any Q&A dataset works. Format:

```json
[
  {"question": "What are the symptoms of diabetes?", "answer": "Diabetes symptoms include..."},
  {"question": "How is hypertension treated?", "answer": "Hypertension treatment involves..."}
]
```

Or use HuggingFace datasets directly:
```bash
python prepare_data.py --source huggingface --dataset keivalya/MedQuad-MedicalQnADataset
```

---

## Quick Start

```bash
cd src/training/domain_adapted_embeddings

# 1. Install dependencies
pip install -r requirements.txt

# 2. Prepare data
python prepare_data.py --source huggingface --dataset keivalya/MedQuad-MedicalQnADataset

# 3. Train
python train.py --data-dir data --output-dir models/trained

# 4. Use trained model
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("models/trained/best", trust_remote_code=True)
embeddings = model.encode(["Your query here"])
```

---

## Limitation: Multi-Domain Integration (TBD)

Each domain-adapted model is fine-tuned independently. When multiple domains exist (medical, legal, finance, etc.), a strategy is needed to integrate them into the semantic router:

| Option | Approach | Trade-off |
|--------|----------|-----------|
| **Multi-domain fine-tuning** | Train one model on all domains together | Simple deployment, but domains may interfere |
| **Two-stage routing** | Domain classifier → domain-specific embeddings | Best accuracy, more complexity |
| **Ensemble** | Run through multiple models, combine scores | Handles cross-domain queries, higher latency |

Currently, this pipeline produces standalone domain models. Integration with the semantic router's existing domain signal infrastructure is future work.

---

## Phase 2: Production Integration (Future)

After POC proves approach works:
1. Load specialized model in Rust via candle
2. FFI bindings for Go
3. Configuration support for domain selection
4. Model registry for multiple domains

---

## References

- Paper: https://arxiv.org/pdf/2512.08088
- Base Model: [llm-semantic-router/mmbert-embed-32k-2d-matryoshka](https://huggingface.co/llm-semantic-router/mmbert-embed-32k-2d-matryoshka)
- Pre-trained Medical Model: [llm-semantic-router/mmbert-embed-medical](https://huggingface.co/llm-semantic-router/mmbert-embed-medical)
- Dataset: [MedQuAD](https://huggingface.co/datasets/keivalya/MedQuad-MedicalQnADataset)

### PR DESCRIPTION
**Pull request by yossiovadia**
Domain-Adapted Embeddings

## Summary

- Implement domain-adapted embedding fine-tuning using iterative hard-negative mining
- Achieve **+71% MRR@5 improvement** on medical domain (MedQuAD dataset)
- Pre-trained model available: [llm-semantic-router/mmbert-embed-medical](https://huggingface.co/llm-semantic-router/mmbert-embed-medical)

Based on: ["Distilling an LLM's Wisdom: A Framework for Creating Domain Adapted Financial Embedding Models"](https://arxiv.org/abs/2512.08088)

## Results

| Metric | Baseline | After Training | Improvement |
|--------|----------|----------------|-------------|
| **MRR@5** | 0.4354 | 0.7453 | **+71.18%** |
| **Recall@5** | 0.4896 | 0.6900 | **+40.92%** |

## Files Added

- `src/training/domain_adapted_embeddings/train.py` - Main training script
- `src/training/domain_adapted_embeddings/prepare_data.py` - Data preparation
- `src/training/domain_adapted_embeddings/README.md` - Overview and results
- `src/training/domain_adapted_embeddings/USAGE.md` - Training instructions

## Test Plan

- [x] Trained on MedQuAD dataset with +71% MRR@5 improvement
- [x] Model uploaded to HuggingFace
- [x] Documentation complete with usage examples

Closes #1131

### CODE REVIEW

**Comment by yossiovadia** [comment]
## Follow-up: Real-World Validation Results

After implementing the finance domain adaptation using the same approach, we conducted real-world validation tests on **both medical and finance models** with out-of-sample Q&A pairs.

### 🔍 Key Findings

**Medical Model Real-World Performance:**
- Tests passed: **2/5 (40%)**
- Average similarity improvement: **-14.2%** (negative)
- Broke 2 working baseline rankings
- Reported training improvement: +71.18% MRR@5

**Finance Model Real-World Performance:**
- Tests passed: **4/5 (80%)**
- Average similarity improvement: **-11.6%** (negative)
- Training improvement: +122.49% MRR@5

### 📊 Pattern Identified

Both models show **strong training metrics but poor real-world generalization**:
- Training datasets: Models learn to rank training examples better
- Real queries: Models perform worse than baseline on new questions
- Overfitting: Training metrics don't predict practical performance

### 🔬 Investigation Status

We are currently investigating:
1. **Training methodology**: Whether iterative hard-negative mining causes overfitting
2. **Evaluation metrics**: If MRR on training data is misleading for domain adaptation
3. **Alternative approaches**: Different fine-tuning strategies that maintain generalization

### 🎯 Recommendations

- **Medical model users**: Test with real medical Q&A before production deployment
- **Approach evaluation**: Consider validation on out-of-sample domain-specific queries
- **Metrics**: Supplement training metrics with real-world validation

### ⚡ Action Items

- [ ] Investigate root causes of training/validation metric disconnect
- [ ] Test alternative domain adaptation approaches  
- [ ] Develop better evaluation protocols for domain-adapted embeddings

Will update as investigation progresses. Thanks for the important work on domain-adapted embeddings! 🧠

**Comment by yossiovadia** [comment]
## LM-Cocktail Solution for Overfitting Issue

Following up on the real-world validation findings, I investigated and implemented a solution based on [LM-Cocktail](https://aclanthology.org/2024.findings-acl.145/) (ACL 2024 Findings).

### Root Cause Analysis

The training is **not teaching domain knowledge** - it's teaching the model to recognize **patterns specific to the training dataset**. This explains why:
- MedQuAD-trained model: +71% on MedQuAD test, but **-14.2% on general medical questions**
- FinanceQA-trained model: +122% on FinanceQA test, but **-16.9% on general finance questions**

The models overfit to dataset patterns (phrasing, structure, vocabulary), not domain expertise.

### Solution: LM-Cocktail Weight Merging

Post-training, merge fine-tuned weights with base model weights:

```python
merged_weights = α × fine_tuned_weights + (1 - α) × base_weights
```

This preserves general knowledge while retaining domain specialization.

### Results

**Finance Domain Test (in-domain):**
| Model | MRR@5 | vs Base |
|-------|-------|---------|
| Fine-tuned (pure) | 0.2960 | +122.49% |
| **Cocktail α=0.7** | **0.3005** | **+125.88%** |

**Real-World Validation (10 out-of-sample Q&A pairs):**
| Model | Pass Rate | vs Base |
|-------|-----------|---------|
| Base | 7/10 (70%) | - |
| Fine-tuned (pure) | **6/10 (60%)** | **-10%** |
| Cocktail α=0.3 | 7/10 (70%) | +0% |
| Cocktail α=0.5 | 7/10 (70%) | +0% |
| **Cocktail α=0.7** | **8/10 (80%)** | **+10%** |

### Key Findings

1. **Pure fine-tuned is WORSE than base** on real-world questions (confirms overfitting)
2. **Cocktail α=0.7 OUTPERFORMS both base AND fine-tuned** on real-world questions
3. Cocktail α=0.7 also achieves **better in-domain performance** (+126% vs +122%)

### Implementation

Added `--lm-cocktail` flag to `train.py`:
```bash
python train.py --data-dir data --output-dir models/trained \
    --lm-cocktail --lm-cocktail-alpha 0.7
```

Standalone script for existing models:
```bash
python apply_cocktail.py \
    --base-model llm-semantic-router/mmbert-embed-32k-2d-matryoshka \
    --finetuned-model models/best \
    --output-dir models/cocktail \
    --alpha 0.7
```

### Recommendations

1. **Always use LM-Cocktail (α=0.7)** as post-processing for domain-adapted models
2. **Validate on out-of-sample questions** before deploying any fine-tuned model
3. **Use diverse training data** if you want true domain expertise (not just dataset pattern matching)

### Files Added
- `apply_cocktail.py` - Standalone weight merging script
- `validate_cocktail_models.py` - Real-world validation script  
- `LM_COCKTAIL_EXPERIMENT.md` - Full documentation

Will submit a follow-up PR with these improvements.

**Comment by yossiovadia** [comment]
## Experiment Results: Dataset Quality & LM-Cocktail Validation

### Key Finding: Dataset Quality Matters More Than Training

We discovered that our original training datasets were too narrow:
- **Medical (MedQuAD)**: Rare disease encyclopedia format ("What is (are) Costeff syndrome?") - Score: 15/100
- **Finance (Doc Extract)**: Invoice extraction patterns ("What is the invoice number?") - Score: 50/100

These caused **catastrophic forgetting** - models improved on training data but got WORSE on real-world questions.

### Better Datasets Identified
| Dataset | Source | Quality Score |
|---------|--------|---------------|
| MedQA (USMLE) | `GBaker/MedQA-USMLE-4-options` | 80/100 |
| FinQA | `virattt/financial-qa-10K` | 70/100 |

### Training Results with Good Datasets
| Model | Training Improvement | Real-World (25 questions) |
|-------|---------------------|---------------------------|
| Medical (MedQA) | +5.69% MRR | 88% (same as base) |
| Finance (FinQA) | +10.08% MRR | 60% (same as base) |
| Combined | +7.21% MRR | 80% med / 60% fin |

### LM-Cocktail Validation: **Confirmed Essential**

Without cocktail, fine-tuned models lose 8% on medical questions. With cocktail (α=0.7), performance is preserved:

| Model | Without Cocktail | With Cocktail |
|-------|-----------------|---------------|
| Medical on medical | 80% | **88%** ✓ |
| Finance on finance | 60% | **60%** ✓ |

### Recommendations
1. **Always use LM-Cocktail** with α=0.7 (now default in `train.py`)
2. **Validate datasets** before training using `validate_dataset.py`
3. **Use specialized models** when domain is known (combined model trades off ~8% on medical)

### New Scripts Added
- `prepare_hf_dataset.py` - Download & prepare HuggingFace datasets
- `validate_dataset.py` - Score dataset quality before training
- `comprehensive_validation.py` - Test models on real-world questions
